serve.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type admitFunc func(v1beta1.AdmissionReview, *config) *v1beta1.AdmissionResponse
+
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc, config *config) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		log.Errorf("contentType=%s, expect application/json", contentType)
+		return
+	}
+
+	//	glog.V(2).Info(fmt.Sprintf("handling request: %v", body))
+	var reviewResponse *v1beta1.AdmissionResponse
+	ar := v1beta1.AdmissionReview{}
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		log.Error(err)
+		reviewResponse = toAdmissionResponse(err)
+	}
+
+	response := v1beta1.AdmissionReview{}
+
+	if ar.Request != nil {
+		reviewResponse = admit(ar, config)
```

Comment:
🤔 so you assign something to reviewResponse on l.38 and then here again? So these two branches are mutually exclusive?

Reply:
They are.. good point, this should be improved upon to be more understandable. I basically copy-pasted this over from a sample and never gave it much thought.